### ISSUE TITLE
fetchSingle() failure does not trigger ExecuteListener#exception() event

### ISSUE THREAD

**Issue by JanecekPetr** [jOOQ/jOOQ#8877]
### Expected behavior and actual behavior:
The `ExecuteListener` contract says:

> An event listener for `Query` [...] fetch steps.

Based on this, I mistakenly expected

```java
return context.select()
    .from(TABLE)
    .where(TABLE.COLUMN.eq(value))
    .fetchSingle()
    .map(MAPPER);
```

to call into my `ExecuteListener.exception()` when the `fetchSingle()` call encounters 0 rows and throws an exception. However, the callback is never called. The same obviously happens when there's too many results, or with the `fetchOne()` call.

Would you please consider changing this so that the execute listeners are called for those specialized `fetch*()` methods, too? We use them almost everywhere and their behaviour is an amazing convenience enforcing that the data layer returns expected results or throws otherwise.

The supplied `ExecuteListener` is, in our case, an exception translator which we rely on to translate jOOQ exceptions into something the rest of the system knows. However, now that we discovered that the `fetchSingle()` thing does not do what we thought it did, we'll probably need to work around it and rewrite our queries to carefully handle row counts explicitly. That's cumbersome, easy to mess up / forget, and forces us to handle exceptions all over the place instead of a single place.

### Steps to reproduce the problem:
Install an `ExecuteListener` with an `exception()` method. Run a query with `fetchSingle()` returning no rows. I really can set up a reproducer project for you, but I think this is so simple that you could manage ;).

### Versions:
- jOOQ: 3.11.11
- Java: 8u201
- Database: Postgres 9.5
- OS: Windows 10
- JDBC Driver: PGJDBC over HikariCP

**Comment by JanecekPetr** [comment]
Oh, I marked it as a bug without thinking too much. Is this behaviour expected? I strongly believe you designed it this way intentionally, so perhaps I should have marked it as a Feature Request. Sorry.

**Comment by lukaseder** [comment]
Thank you very much for your report.

I'm undecided whether this is a bug or "feature", although I agree these exceptions should reach the `ExecuteListener`, as they are happening during the execution lifecycle, even if from outside the internal `Cursor` instance. All of these calls are wired to `Tools.fetchOne()` or `Tools.fetchSingle()` methods, which manage the `Cursor` lifecycle through `cursor.close()` calls. Exceptions are thus throws prior to the `ExecuteListener.end()` event:

https://github.com/jOOQ/jOOQ/blob/version-3.11.11/jOOQ/src/main/java/org/jooq/impl/Tools.java#L1853-L1873

While I agree this should be fixed, it might constitute an incompatible change of behaviour, so we cannot backport this fix. We'll implement it in 3.12, though.

**Comment by lukaseder** [comment]
In fact, this usage of jOOQ API will also prevent `ExecuteListener#exception()` calls when something goes wrong while iterating the JDBC `ResultSet`, e.g. when a `ResultSet.getString()` method call fails.

**Comment by lukaseder** [comment]
My bad. Exceptions thrown from `ResultSet` operations do pass by `ExecuteListener#exception()`

**Comment by lukaseder** [comment]
Related issue: #8544

**Comment by lukaseder** [comment]
> However, now that we discovered that the fetchSingle() thing does not do what we thought it did, we'll probably need to work around it and rewrite our queries to carefully handle row counts explicitly.

As a workaround, you could instrument the jOOQ API to catch all `fetchOne()` and `fetchSingle()` calls, put this fact in some thread local variable, and then use an `ExecuteListener` that checks the row count explicitly on the `fetchEnd()` or `end()` event, and throws your custom exception if it is wrong.

**Comment by lukaseder** [comment]
This first commit fixes the issue for the most common `fetchOne()` and `fetchSingle()` calls. There are others, that use the `Tools.filterOne()` method, which might still be wrong. Also, the order of callback calls tends to be wrong because of #8544, which is probably not relevant to your use-case, but should still be considered in the same go.

**Comment by lukaseder** [comment]
A separate discussion could be had about optimistic locking exceptions, and other `UpdatableRecord` related exceptions, which are less obvious bug/feature request decisions. I'll leave them out of scope for now: #8880

**Comment by lukaseder** [comment]
Another out-of-scope, but related issue is the lack of `ExecuteListener#exception()` lifecycle event when using `fetchOne()` on DML `RETURNING` clauses: #8881. This will be fixed in a future release.

**Comment by lukaseder** [comment]
Fixed in 3.12. I will not backport this fix as it could lead to very subtle regressions in client code that would not expect this exception to be part of the `ExecuteListener` lifecycle.
